cobrawap/pipeline/stage03_trigger_detection/scripts/minima.py

```diff
@@ -56,36 +56,50 @@ def filter_minima_order(signal, mins, order=1):
     return filtered_mins
 
 
-def moving_threshold(signal, window, fraction):
-    # compute a dynamic threshold function through a sliding window 
-    # on the signal array
-    strides = np.lib.stride_tricks.sliding_window_view(signal, window)
-    threshold_func = np.min(strides, axis=1) + fraction*np.ptp(strides, axis=1)
+def moving_threshold(signal, maxima_threshold_window, maxima_threshold_fraction):
+    threshold_signal = []
+
+    sampling_rate = signal.sampling_rate.rescale('Hz').magnitude
+    duration = float(signal.t_stop.rescale('s').magnitude - signal.t_start.rescale('s').magnitude)
+    if maxima_threshold_window is None or maxima_threshold_window > duration:
+        maxima_threshold_window = duration
+    window_frame = int(maxima_threshold_window*sampling_rate)
+
+    for channel, channel_signal in enumerate(signal.T):
+        if np.isnan(channel_signal).any():
+            threshold_func = np.full(np.shape(signal)[0], np.nan)
+        else:
+            # compute a dynamic threshold function through a sliding window
+            # on the signal array
+            strides = np.lib.stride_tricks.sliding_window_view(channel_signal, window_frame)
+            threshold_func = np.min(strides, axis=1) + maxima_threshold_fraction*np.ptp(strides, axis=1)
+            # add elements at the beginning
+            threshold_func = np.append(np.ones(window_frame//2)*threshold_func[0], threshold_func)
+            threshold_func = np.append(threshold_func, np.ones(len(channel_signal)-len(threshold_func))*threshold_func[-1])
+        threshold_signal.append(threshold_func)
+
+    threshold_signal = neo.AnalogSignal(np.array(threshold_signal).T, units=signal.units, sampling_rate=signal.sampling_rate)
```

Comment:
Why create a separate threshold analogsignal here? It seems it requires more code and more RAM

Reply:
In the previous version of the code, the threshold was computed channel-wise in the `detect_minima` function, and then again in `plot_minima` function for each channel to be plotted. The new implementation does it only once for each channel.
Also, for a given channel to be visualized, the signal to be plotted was trimmed as [plot_tstart, plot_tstop] before passing it to `plot_minima` function. This implied that the moving threshold had to be computed on the signal after the trimming, with minima found and plotted consequently. This was different from what done in `detect_minima` function, where moving thresholds and minima were computed on the whole signal, with no trimming. This means that minima reported in plots could have been different from minima actually detected for a certain channel. Thanks to @mdenker for pointing this out during one of our discussions.
So, we decided to coherently compute moving threshold on the whole signal for each channel, only once, and then use such `threshold_signal` for coming minima detection and visualization. On one side, this certainly implies a larger RAM usage, but ensures a more coherent and robust computation of the moving thresholds. At the same time, larger number of code lines in `moving_threshold` function are almost balanced by code lines now useless and hence removed elsewhere in the code.